### ISSUE TITLE
Release DataFusion `44.0.0`

### ISSUE THREAD

**Issue by alamb** [apache/datafusion#13334]
### Is your feature request related to a problem or challenge?

Tracking ticket for next release, also a place to track desired inclusions

Last release was https://crates.io/crates/datafusion/43.0.0 November 8th, 2024 so next major release would be around December 8, 2024

* Voting Thread: https://lists.apache.org/thread/78zxfpn7fh1csz6fdolpgh3lb4ltxngp

Prior release tickets:
*  43.0.0 https://github.com/apache/datafusion/issues/12470

## Items to possibly fix before release

- [ ] https://github.com/apache/datafusion/issues/13510
- [ ] https://github.com/apache/datafusion/issues/13504

Improved upgrade experience:
- [x] https://github.com/apache/datafusion/issues/13763
- [x] https://github.com/apache/datafusion/issues/13764
- [x] https://github.com/apache/datafusion/issues/13766
- [ ] https://github.com/apache/datafusion/pull/13717

Downstream project testing: 
- [x] datafusion python: https://github.com/apache/datafusion-python/issues/972
- [x] comet: https://github.com/apache/datafusion/issues/13835
- [x] delta.rs: https://github.com/apache/datafusion/issues/13834
- [x] https://github.com/apache/datafusion/issues/13855

**Comment by alamb** [comment]
I would love to get this one in:
- https://github.com/apache/datafusion/issues/13516

Some things to highlight (that are already merged)
- https://github.com/apache/datafusion/pull/12996 from @Rachelint  as well

**Comment by alamb** [comment]
@Omega359 and @andygrove suggested https://github.com/apache/datafusion/issues/13525#issuecomment-2496487538 that for this release we 

> we file tickets as part of every release to test DF with latest versions of common external dependencies (datafusion-python, iceberg, lance, delta, etc) as part of the gating criteria for a release. Perhaps this would be as simple as changing the DF version in those projects and running their test suite. Any blockers that are found in DF have tickets filed and are fixed, other blockers in the respective projects have tickets filed in those projects


Perhaps for 44.0.0 we can try with own subprojects (Ballista, Comet, DF Python, DF Ray)

**Comment by alamb** [comment]
I started gathering a list of items we think we should fix prior to the release in the description

**Comment by alamb** [comment]
Does anyone have any opinion about holding the DataFusion 44 release for the next major arrow release?
- https://github.com/apache/arrow-rs/issues/6342

That would fix a bunch of the StringView issues, but would likely delay DataFusion 44 for a few more weeks

**Comment by Omega359** [comment]
It's soon to be the holiday season so I'm all for cooking the release a bit more.

**Comment by alamb** [comment]
I would personally love to see DataFusion 44.0.0 be lauded as "super stable" and have few upgrade issues (we would largely achieve this by testing upgrades with other projects prior to release)

The downside is that the changeset in DataFusion 44.0.0 might be larger

**Comment by andygrove** [comment]
I'd like to see an upgrade guide for the 44 release (and am willing to take the lead on this). I am trying to upgrade Comet now and am running into some issues.

I am +1 for waiting for the next Arrow release.

**Comment by andygrove** [comment]
I filed https://github.com/apache/datafusion/issues/13702

**Comment by andygrove** [comment]
How would everyone feel about increasing the length of release votes from 3 days to 7 days to give downstream projects more time to test the release?

Sometimes the vote starts on a Friday and passes on a Monday and we can't expect everyone to be working weekends.

**Comment by alamb** [comment]
> How would everyone feel about increasing the length of release votes from 3 days to 7 days to give downstream projects more time to test the release?
> 
> Sometimes the vote starts on a Friday and passes on a Monday and we can't expect everyone to be working weekends.

In my opinion, we would ideally do the testing before we make the release candidate (as the overhead of making RCs is non trivial). However, I am not opposed to extending the voting timeline if that would get us more testing time

**Comment by andygrove** [comment]
Testing before we create the RC makes sense

**Comment by andygrove** [comment]
I'm still working on updating Comet to work with latest DF from main (upgrading from 43.0.0). This is quite a disruptive upgrade for us so far.

![Screenshot from 2024-12-09 17-37-14](https://github.com/user-attachments/assets/91a653fd-25a5-4494-87ac-52b09a3c9a2f)

**Comment by alamb** [comment]
I have been thinking about this upgrade and the next arrow release
- https://github.com/apache/arrow-rs/issues/6342

Specifically I think we need this upgrade in order to resolve some of the issues in 44 related to string view
- [ ] https://github.com/apache/datafusion/issues/13504

However, what I would like to propose is:
1. Plan to release DataFusion 44.0.0 *without* also upgrading arrow-rs

The downside of this plan is
1.  it will take longer to get the arrow upgrade (and this the string view fixes) throughout the ecosystem
The upsides are 
2. It will reduce the complexity of the DataFusion 44 upgrade
3. it will give us a longer period to test the upgrade integration DataFusion

If someone is critically waiting on the string view fixes, we can also contemplate making another arrow incremental release

**Comment by alamb** [comment]
I am also going to coordinate with the delta-rs folks to try and test upgrading prior to release:
- https://github.com/delta-io/delta-rs/pull/2886#issuecomment-2541251732

**Comment by timsaucer** [comment]
Starting an issue and branch in `datafusion-python` : https://github.com/apache/datafusion-python/issues/972

**Comment by rluvaton** [comment]
I'll give my 2 cents from my experience regarding how other popular projects handle this issue

In Node.js (I'm a core collaborator) before each release we run the tests of the most popular packages with the new node version and making sure we are not breaking them, if we are we let them know. this is really helpful to avoid breaking the entire world

this process is a [CITGM](https://github.com/nodejs/citgm), you can watch this [talk](https://www.youtube.com/watch?v=8is8iKlo8oQ) for more info

**Comment by andygrove** [comment]
I am +1 for releasing DF 44 without the next major version of Arrow. However, there are some bug fixes and performance improvements in Arrow that I would like to use so would like to see a new minor release (and am happy to help with this if needed).

**Comment by mbwhite** [comment]
Hi - would like to get #13647 out as soon as is practical - a small fix but does unblock my use of 
DataFusion for Substrait.

Thanks!

**Comment by alamb** [comment]
> I am +1 for releasing DF 44 without the next major version of Arrow. However, there are some bug fixes and performance improvements in Arrow that I would like to use so would like to see a new minor release (and am happy to help with this if needed).

- I have filed https://github.com/apache/arrow-rs/issues/6887 to track this potential release

**Comment by alamb** [comment]
I think we should start pushing to make this release before it accumulates more API changes such as
- https://github.com/apache/datafusion/pull/13823

@andygrove are you willing to try and make this release later this week? Maybe we can rally people to agree on what we need for the next release and try and get the upgrades lines up

**Comment by Blizzara** [comment]
(Just for what it's worth, I appreciate smaller releases more often compared to bigger ones less often, even if it means more more times I need to fix downstream stuff due to breaks, presumably it usually also means less big breaks. And also faster time to get my and other's improvements and fixes into test/use, and thus faster finding of unintended regressions when they happen. That said, I appreciate there's a bunch of factors including the work needed etc, just wanted to offer my 2 cents 😄)

**Comment by andygrove** [comment]
> @andygrove are you willing to try and make this release later this week? Maybe we can rally people to agree on what we need for the next release and try and get the upgrades lines up

@alamb Sure, but I didn't have time yet to help with creating a migration guide, other than the notes I made on https://github.com/apache/datafusion/issues/13702

**Comment by andygrove** [comment]
The main thing we are waiting on for Comet is https://github.com/apache/datafusion/pull/13778

**Comment by alamb** [comment]
> Just for what it's worth, I appreciate smaller releases more often compared to bigger ones less often, even if it means more more times I need to fix downstream stuff due to breaks, presumably it usually also means less big breaks. And also faster time to get my and other's improvements and fixes into test/use, and thus faster finding of unintended regressions when they happen. That said, I appreciate there's a bunch of factors including the work needed etc, just wanted to offer my 2 cents 😄)

I agree -- thank you @Blizzara 

In this case I think we are running into bandwidth / maintainer limitations -- I would really like to make sure our releases are less jarring

Does anyone have time to help test with delta.rs?
-  https://github.com/apache/datafusion/issues/13834

**Comment by findepi** [comment]
@shehabgamin that would be awesome, thanks!
It could also help if we could land some of these tests in datafusion itself. This would help avoid breakages before a PR is merged.

**Comment by ozankabak** [comment]
> It could also help if we could land some of these tests in datafusion itself. This would help avoid breakages before a PR is merged.

Agreed, let's do this

**Comment by alamb** [comment]
> We'll integrate Sail with the latest commit in the DataFusion main branch and report back our findings. When should Sail complete testing by? We do have bandwidth this week and next week. We’d be happy to prioritize testing this on our side, just lemme know expectations around timing!

I am personally hoping we are ready to make an RC from main by early next week

> > It could also help if we could land some of these tests in datafusion itself. This would help avoid breakages before a PR is merged.
> 
> Agreed, let's do this

100% agree -- a good place to start might be any tests that you find fail in your project but pass all DataFusion tests

**Comment by andygrove** [comment]
> I am personally hoping we are ready to make an RC from main by early next week

Just fyi, I may not be online much next week, not that I need to be involved in cutting the RC, but will likely have time the following week to help, if needed.

**Comment by alamb** [comment]
What I plan to do is try and clean up any remaining outstanding PRs and get them merged over the next few days and I'll try and prepare the RC asap

Any help people can give to help figure out how to get the approved PRs merged would be super helpful: 
- https://github.com/apache/datafusion/pulls?q=is%3Apr+is%3Aopen+review%3Aapproved+-is%3Adraft

**Comment by alamb** [comment]
Status update here:

## Ease upgrades:
Summary: From my perspective the following issues are nice to have, rather than required for the release
- [ ] https://github.com/apache/datafusion/issues/13763 / https://github.com/apache/datafusion/pull/13881
- [ ] https://github.com/apache/datafusion/issues/13764 / https://github.com/apache/datafusion/pull/13870


Pre-release testing:
- [x] Comet upgrade looks good, thanks to @andygrove: https://github.com/apache/datafusion-comet/pull/1154#issuecomment-2558475306
- [x] Delta-rs upgrade looks good, thanks to @rtyler and @ion-elgreco:  https://github.com/delta-io/delta-rs/pull/3073
- [ ] datafusion-python looks good, thanks to @timsaucer : https://github.com/apache/datafusion-python/pull/973

Other items:
- [ ] Needs triage (maybe is already fixed): https://github.com/apache/datafusion/issues/13510

## StringView support
These are mostly waiting on arrow-rs updates (I think we can fix many of them with the [arrow-rs 53.4.0](https://github.com/apache/arrow-rs/issues/6887) release
- [ ]  https://github.com/apache/datafusion/issues/13504

**Comment by alamb** [comment]
I am looking at https://github.com/apache/datafusion/issues/13510 now, and then plan to make a PR with release notes, verison udate, etc

**Comment by andygrove** [comment]
@alamb The miri issue does not seem to be resolved for Comet - https://github.com/apache/datafusion/issues/13835#issuecomment-2558543030

Is there anything I should have done other than set `default-features=false` on all DF crates?

**Comment by alamb** [comment]
> @alamb The miri issue does not seem to be resolved for Comet - [#13835 (comment)](https://github.com/apache/datafusion/issues/13835#issuecomment-2558543030)
> 
> Is there anything I should have done other than set `default-features=false` on all DF crates?

That sounds like it is the right thing to me 🤔 

- Since @buraksenn did the work to make this optional in  https://github.com/apache/datafusion/pull/13778, perhaps they have some idea?

**Comment by alamb** [comment]
> Since @buraksenn did the work to make this optional in [minor] make recursive package dependency optional  #13778, perhaps they have some idea?

Update I think I know what is wrong -- working on a fix now

**Comment by alamb** [comment]
I think all we are waiting on now is someone to review / approve (or fix in some other way) the `recrusive-protection` feature:
- https://github.com/apache/datafusion/pull/13887

**Comment by alamb** [comment]
Update, when preparing to make a final RC I found another issue:
- https://github.com/apache/datafusion/issues/13898

I have a proposed fix here:
- https://github.com/apache/datafusion/pull/13899

**Comment by alamb** [comment]
Ok, we are clear for 🚀 from what I can tell -- the comet upgrade is working. I will merge the following PR
- https://github.com/apache/datafusion/pull/13882 (thanks @xudong963 and @phillipleblanc   for the review 🙏 )
And then make a RC

**Comment by alamb** [comment]
I made a release candidate: https://lists.apache.org/thread/78zxfpn7fh1csz6fdolpgh3lb4ltxngp

**Comment by alamb** [comment]
Thanks @shehabgamin

**Comment by timsaucer** [comment]
If we can make a new RC I would recommend cherry picking this fix for initcap: https://github.com/apache/datafusion/pull/13909

**Comment by alamb** [comment]
As @timsaucer  mentions above (and on the mailing list): https://lists.apache.org/thread/34xy1nm47tkrr8lszxzwcrlf5b47h333

There was a regression introduced into initcap that has been sinced fixed: https://github.com/apache/datafusion/pull/13909

I will make a new RC

**Comment by alamb** [comment]
have created a new RC and started a new vote thread: https://lists.apache.org/thread/p0hyr78zs03f3snj0mh3vqm6c3kjwshn

**Comment by alamb** [comment]
The release has been approved and published to crates.io: https://lists.apache.org/thread/7w64xqzn71plvmxy61v6px8bb994kr79 🎉 

https://crates.io/crates/datafusion/44.0.0

I have also made a small PR to improve the docs for the next release:
- https://github.com/apache/datafusion/pull/13959

**Comment by niebayes** [comment]
@alamb Hi, is the release for datafusion-cli delayed?

**Comment by alamb** [comment]
> @alamb Hi, is the release for datafusion-cli delayed?

Good catch @niebayes  -- I forgot to push this to crates.io. I have done so now and `datafusion-cli` 44.0.0 is available:

https://crates.io/crates/datafusion-cli/44.0.0

**Comment by alamb** [comment]
Here is a ticket to run ClickBench on DataFusion 44:
- https://github.com/apache/datafusion/issues/13983

**Comment by alamb** [comment]
Filed ticket for the next relese:
- https://github.com/apache/datafusion/issues/14008
